README.md

```diff
@@ -649,12 +649,406 @@ To begin coding the basic execution functionality of our VM, we only need these
+Now its time to decode the instruction and then run the `decoded instruction`.
 
-#### Run instruction
 
 ```go
-    //TODO
+func (v *VirtualMachine) Step() error {
+	encoded_instruction, err := v.Segments.Memory.Get(v.RunContext.Pc)
+	if err != nil {
+		return fmt.Errorf("Failed to fetch instruction at %+v", v.RunContext.Pc)
+	}
+
+	encoded_instruction_felt, ok := encoded_instruction.GetFelt()
+	if !ok {
+		return errors.New("Wrong instruction encoding")
+	}
+
+	encoded_instruction_uint, err := encoded_instruction_felt.ToU64()
+	if err != nil {
+		return err
+	}
+
+	instruction, err := DecodeInstruction(encoded_instruction_uint)
+	if err != nil {
+		return err
+	}
+
+	return v.RunInstruction(&instruction)
+}
+```
+
+##### Decode instruction 
+
+```go
+//  Structure of the 63-bit that form the first word of each instruction.
+//  See Cairo whitepaper, page 32 - https://eprint.iacr.org/2021/1063.pdf.
+// ┌─────────────────────────────────────────────────────────────────────────┐
+// │                     off_dst (biased representation)                     │
+// ├─────────────────────────────────────────────────────────────────────────┤
+// │                     off_op0 (biased representation)                     │
+// ├─────────────────────────────────────────────────────────────────────────┤
+// │                     off_op1 (biased representation)                     │
+// ├─────┬─────┬───────┬───────┬───────────┬────────┬───────────────────┬────┤
+// │ dst │ op0 │  op1  │  res  │    pc     │   ap   │      opcode       │ 0  │
+// │ reg │ reg │  src  │ logic │  update   │ update │                   │    │
+// ├─────┼─────┼───┬───┼───┬───┼───┬───┬───┼───┬────┼────┬────┬────┬────┼────┤
+// │  0  │  1  │ 2 │ 3 │ 4 │ 5 │ 6 │ 7 │ 8 │ 9 │ 10 │ 11 │ 12 │ 13 │ 14 │ 15 │
+// └─────┴─────┴───┴───┴───┴───┴───┴───┴───┴───┴────┴────┴────┴────┴────┴────┘
+```
+
+As we can see in the chart above, all the information we need is present on the bits representation of the instruction. The first thing to do is create a structure that stores it. 
+
+```go 
+type Instruction struct {
+	Off0     int
+	Off1     int
+	Off2     int
+	DstReg   Register
+	Op0Reg   Register
+	Op1Addr  Op1Src
+	ResLogic ResLogic
+	PcUpdate PcUpdate
+	ApUpdate ApUpdate
+	FpUpdate FpUpdate
+	Opcode   Opcode
+}
+```
+
+And the good thing about this, is every bits combination for each field is known, so we can code all the possible flags to work with. These flags are represented below. 
+
+The Register constants, there's two possible registers, ap and fp. The ap register (address pointer register) keeps track of memory addresses for data access. The fp register (frame pointer register) manages function call stack frames, local variables and parameter access.
+
+```go 
+// x-----------------------------x
+// x----- Instruction flags -----x
+// x-----------------------------x
+
```

Comment:
We should also mention the offsets `Off0, Off1, Off2` what they mean, and which operand each one affects, We could explain here that we use the offset and the register to obtain the value of each operand, for example if the DstReg is AP and Off0 is -1, then the dst operand will be located one position before the current AP value

Reply:
I would add this on the compute operands documentation, because is when they are used. I've done a draft version of what I would say, maybe it can be useful. What do you think? 
```suggestion
The Off0 is used to compute the result address of the instruction paired with the destination register. It can be a negative number, so if Off0 is -1 and the dstreg is ap, then, the result of the instruction will be loaded in one position before of the address pointed by ap. 

    The Off1 is used to compute the first operand (op0) address of the instruction paired with the op0 register. It can be a negative number too. 

    Finally the Off2, is used to compute the second operand (op1) address, 
    this one is a little bit special because it also can be used with the pc register when its value is 1. A part from that, it works the same as the other offsets. 

```
I mention @mmsc2 in the case is better to document in that part.

Reply:
Maybe we can do something shorter, for example:
```
The off0, off1, and off2 values are used to compute the address of the dst, op0 and op1 respectively.
For example, if the DstReg is AP and Off0 is -1, then we can compute the dst address by substracting one from the current value of ap.
```
Its better to go more in detail in the compute_operands section but we should also give a brief description here so that the reader is not left wandering what the offsets in the instruction mean

Reply:
I agree, better to document too much than too little. Thank you so much!